tests/legacy-cli/e2e/utils/process.ts

```diff
@@ -69,48 +67,65 @@ function _exec(options: ExecOptions, cmd: string, args: string[]): Promise<Proce
   }
 
   const childProcess = child_process.spawn(cmd, args, spawnOptions);
-  childProcess.stdout!.on('data', (data: Buffer) => {
-    stdout += data.toString('utf-8');
-    if (options.silent) {
-      return;
-    }
-    data
-      .toString('utf-8')
-      .split(/[\n\r]+/)
-      .filter((line) => line !== '')
-      .forEach((line) => console.log('  ' + line));
-  });
-
-  childProcess.stderr!.on('data', (data: Buffer) => {
-    stderr += data.toString('utf-8');
-    if (options.silent) {
-      return;
-    }
-    data
-      .toString('utf-8')
-      .split(/[\n\r]+/)
-      .filter((line) => line !== '')
-      .forEach((line) => console.error(colors.yellow('  ' + line)));
-  });
 
   _processes.push(childProcess);
 
   // Create the error here so the stack shows who called this function.
   const error = new Error();
 
-  // Return log info about the current process status
-  function envDump() {
-    return [
-      `ENV:${JSON.stringify(spawnOptions.env, null, 2)}`,
-      `STDOUT:\n${stdout}`,
-      `STDERR:\n${stderr}`,
-    ].join('\n\n');
-  }
-
   return new Promise<ProcessOutput>((resolve, reject) => {
+    let stdout = '';
+    let stderr = '';
     let matched = false;
 
-    childProcess.on('exit', (code: number) => {
+    // Return log info about the current process status
+    function envDump() {
+      return [
+        `ENV:${JSON.stringify(spawnOptions.env, null, 2)}`,
+        `STDOUT:\n${stdout}`,
+        `STDERR:\n${stderr}`,
+      ].join('\n\n');
+    }
+
+    childProcess.stdout!.on('data', (data: Buffer) => {
+      stdout += data.toString('utf-8');
+
+      if (options.waitForMatch && stdout.match(options.waitForMatch)) {
+        resolve({ stdout, stderr });
+        matched = true;
+      }
+
+      if (options.silent) {
+        return;
+      }
+
+      data
+        .toString('utf-8')
+        .split(/[\n\r]+/)
+        .filter((line) => line !== '')
+        .forEach((line) => console.log('  ' + line));
+    });
+
+    childProcess.stderr!.on('data', (data: Buffer) => {
+      stderr += data.toString('utf-8');
+
+      if (options.waitForMatch && stderr.match(options.waitForMatch)) {
+        resolve({ stdout, stderr });
+        matched = true;
+      }
+
+      if (options.silent) {
+        return;
+      }
+
+      data
+        .toString('utf-8')
+        .split(/[\n\r]+/)
+        .filter((line) => line !== '')
+        .forEach((line) => console.error(colors.yellow('  ' + line)));
+    });
+
+    childProcess.on('close', (code: number) => {
```

Comment:
Note this changed from `exit` to `close` to ensure stdout/err have completed:

https://nodejs.org/api/child_process.html#event-close
> The 'close' event is emitted after a process has ended and the stdio streams of a child process have been closed

https://nodejs.org/api/child_process.html#event-exit
> When the 'exit' event is triggered, child process stdio streams might still be open.